src/neo4j_genai/retrievers/external/pinecone/pinecone.py

```diff
@@ -0,0 +1,187 @@
+#  Copyright (c) "Neo4j"
+#  Neo4j Sweden AB [https://neo4j.com]
+#  #
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#  #
+#      https://www.apache.org/licenses/LICENSE-2.0
+#  #
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import logging
+from typing import Any, Callable, Optional
+
+import neo4j
+from pinecone import Pinecone
+from pydantic import ValidationError
+
+from neo4j_genai.embedder import Embedder
+from neo4j_genai.exceptions import (
+    EmbeddingRequiredError,
+    RetrieverInitializationError,
+    SearchValidationError,
+)
+from neo4j_genai.retrievers.base import ExternalRetriever
+from neo4j_genai.retrievers.external.pinecone.types import (
+    PineconeClientModel,
+    PineconeNeo4jRetrieverModel,
+    PineconeSearchModel,
+)
+from neo4j_genai.retrievers.external.utils import get_match_query
+from neo4j_genai.types import (
+    EmbedderModel,
+    Neo4jDriverModel,
+    RawSearchResult,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class PineconeNeo4jRetriever(ExternalRetriever):
+    def __init__(
+        self,
+        driver: neo4j.Driver,
+        client: Pinecone,
+        index_name: str,
+        id_property_neo4j: str,
+        embedder: Optional[Embedder] = None,
+        return_properties: Optional[list[str]] = None,
+        retrieval_query: Optional[str] = None,
+        format_record_function: Optional[Callable] = None,
```

Comment:
Is this being used?

Reply:
It should be added to the instance:
```
        self.format_record_function = validated_data.format_record_function
```
and will be used by the `search` method automatically if not None.

Reply:
`format_record_function`? Not currently. I saw it was added to the Weaviate retriever though (where it also doesn't appear to be doing anything) and assumed it was something we would looking to implement something using it in the future. Happy to remove it though if it's not needed.

Reply:
See https://github.com/neo4j/neo4j-genai-python/blob/main/src/neo4j_genai/retrievers/external/weaviate/weaviate.py#L81 (what Estelle said)

Reply:
Cool, I've updated it, should be good to go now I believe.